### ISSUE TITLE
Proposal: New module structure

### ISSUE THREAD

**Issue by kirjs** [kirjs/react-highcharts#41]
# Problem 
Due to the complex structure of `Highcharts` there has been a lot of confusion around how to get the right combination of `Highcharts` modules into the codebase. 

Also for every additional module or variation of `Highcharts` a separate pull request has to be created, each of them following some different structure, making using/contributing to `react-highcharts` more and more confusing.

Issues #18, #32, #33, #38, #39 and #40 are still open because of that.

# Proposed solution
Below I propose the new module structure, which would hopefully bring in understandable and predictable module structure, and will make `react-highcharts` easier to use and to contribute to.

## Default
`require('react-highcharts')`
By default only the `React` component should be required, expecting that `Highcharts` was required independently, and a global variable exists.

This is a breaking change, and addresses #40. 

## Bundles
`require('react-highcharts/bundles/highcharts'); // Acts as require('react-highcharts') now` 
`require('react-highcharts/bundles/highmaps');`
`require('react-highcharts/bundles/highstock');`
Bundled versions come with their own version of Highcharts/Highmaps/Highstocks/Highwhatever. 

They are meant to simplify the usage of the component, and no to bundles can be loaded simultaneously, as that would bring 2 versions of the highcharts into the codebase. 

This would address #18

## Modules
`require('react-highcharts/modules/more')`;
`require('react-highcharts/modules/exporting')`;
`require('react-highcharts/modules/heatmap')`;
etc. 

Modules will be used as a shortcuts to the modules folder in `highcharts-release`. 
They won't have `Highcharts` bundled, and can be required in any combinations.

This would address #32, #33, #38, #39,

**Comment by kirjs** [comment]
cc @aldendaniels @Kureev @joshhornby

**Comment by Kureev** [comment]
why not just `react-highcharts/<name-of-module>`?

**Comment by kirjs** [comment]
1) So that it's explicit that those are modules
2) So that we don't have a bunch of files in the roue. 
3) I was also thinking of generating them somehow, and this way files would be easier to clean and deal with if they are in a separate folder

**Comment by Kureev** [comment]
If I'm right, you can point `index` file in package.json in `lib` folder (or whatever) and in the same folder store all your module files, in this case relative path would be calculated relatively to index.

**Comment by Kureev** [comment]
Or, of course, you can create absolutely different packages and use them as plugins

**Comment by aldendaniels** [comment]
A similar (but modified) idea...

## Idea

1. Publish standalone NPM modules for the various Highcharts flavors we want to support with properly defined exports. Of course Highcharts really should be doing this for us... but they aren't. :cry: 

2. Add an `injectHighcharts()` method to React Highcharts to get a handle to whatever copy of Highcharts the user is using.

3. Stop bundling Highcharts.

4. Then users can do this:
    ```javascript
    import Highcharts from 'my-flavor-of-highcharts';
    import ReactHighcharts from 'react-highcharts';
    ReactHighcharts.injectHighcharts(Highcharts);
    
    // Then use <ReactHighcharts /> like normal.
```

## Benefits

 1. We can separately version the **Highchart modules** and **React Highcharts** itself. This means that a breaking Highcharts change won't require a major version change for React Highcharts.

 3. We can separately version each Highcharts module. This is important if we want to support other Highcharts libraries such as Highmaps.

 2. Non-React developers can use the Highcharts NPM modules we'd be maintaining. This is good for everyone.

 3. The injector method avoids reliance on a global Highcharts variable. This reduces the risk of order-dependent module loading problems.

**Comment by aldendaniels** [comment]
@Kureev - Just saw your comment:

> Or, of course, you can create absolutely different packages and use them as plugins

I think we're getting at the same idea.

**Comment by kirjs** [comment]
## Default
`var ReactHighcharts = require('react-highcharts')` will expect a global Highcharts variable
Optionally it would be possible to inject a different version of highcharts
`import Highcharts from 'my-flavor-of-highcharts';
import ReactHighcharts from 'react-highcharts';
ReactHighchartsWithFlavor = ReactHighcharts.injectHighcharts(Highcharts);`
`injectHighcharts` in this case acts as a factory and creates a copy of reactHighcharts using different version of Highcharts, so it's possible to have several versions of react Highcharts if necessary. 

## Bundles
Looks like everybody agrees on bundles

## Modules
Looks like most the ideal solution would be to have highcharts org to have each module as a separate npm package. 
I actually see better modularization in their roadmap for highcharts 5. 
http://www.highcharts.com/support/roadmap

This is coming on Q1 though, and as a temporary solution we could follow @Kureev's suggestion and do something like this:

`require('react-highcharts/more')`;
`require('react-highcharts/exporting')`;
`require('react-highcharts/heatmap')`;

**Comment by adamshone** [comment]
+1 for @Kureev's suggestion for modules. I'm trying (and failing) to get drilldown (http://www.highcharts.com/docs/chart-concepts/drilldown) working via react-highcharts at the moment, `require('react-highcharts/drilldown');` sounds ideal.

**Comment by aldendaniels** [comment]
@kirjs - Looking good. I'd still suggest moving the Highcharts builds into separate NPM packages:

 - `highcharts-build-more`
 - `highcharts-build-exporting`
 - `highcharts-build-heatmap`

This is much more friendly to non-react consumers and makes for better semserver.

**Comment by kirjs** [comment]
@aldendaniels Will move out the modules once #44 is done

**Comment by kirjs** [comment]
Should be fixed starting with v3
